scripts/dump_state_dict.py

```diff
@@ -126,13 +119,9 @@ def group_keys(state: State, level: int = 0) -> Fork[State] | State:
             return state
         current_keys.append(parts[level])
 
-    by_key: dict[str, State] = {}
-    for i, (k, v) in enumerate(state.items()):
-        key = current_keys[i]
-        if key not in by_key:
-            by_key[key] = {}
-        key_state = by_key[key]
-        key_state[k] = v
+    by_key: Mapping[str, State] = collections.defaultdict(dict)
+    for key, (k, v) in zip(current_keys, state.items()):
+        by_key[key][k] = v
```

Comment:
Nothing about your code, but I want to rant about `collections.defaultdict`:

Which one of the geniuses at python HQ that adding a `get_or_add(key: K, supplier: Callable[[], V]) -> V` method ([or](https://www.baeldung.com/java-map-computeifabsent) [similar](https://doc.rust-lang.org/std/collections/hash_map/enum.Entry.html#method.or_default)) method to `dict` was too easy? Instead, they gave us a subtype of `dict` that modifies itself on reads. [`setdefault`](https://docs.python.org/3/library/stdtypes.html#dict.setdefault) was a good step, but it apparently never occurred to them that constructing a default value every time might not be ideal.

`dict` is one of the most fundamental and most used container types in python, so how can its API be so utterly simplistic? Even Java added this stuff in Java 8 10 years ago. It's almost as bad as JS' `Map` class with the difference being that `dict` is way older.